File: topics/class/fractions/att/template.py
#!/usr/bin/env python3


class Fraction:
    """
    A number represented as a fraction
    """

    def __init__(self, num, denom):
        """num and denom!=0 are integers
        call reduce here"""
        # Write your code here

    # Getters
    def get_num(self):
        return self.__num

    def get_denom(self):
        return self.__denom

    # Setters
    def set_num(self, value):
        assert type(value) == int, "numerator must be an int"
        self.__num = value
        self.reduce()

    def set_denom(self, value):
        assert type(value) == int, "denominator must be an int"
        assert value != 0, "denominator cannot be zero"
        self.__denom = value
        self.reduce()

    def reduce(self):
        """Reduce fraction to lowest terms
        Keep denominator positive"""
        # Write your code here

    def __str__(self):
        """Returns a string representation of self
        For example '-1/2' if numerator = -1 and denominator = 2"""
        # Write your code here

    def __add__(self, other):
        """Returns a new fraction representing the addition"""
        # Write your code here

    def __sub__(self, other):
        """Returns a new fraction representing the subtraction"""
        # Write your code here

    def __mul__(self, other):
        """Returns a new fraction representing the multiplication"""
        # Write your code here

    def __truediv__(self, other):
        """Returns a new fraction representing the division"""
        # Write your code here

    def __float__(self):
        """Returns a float value of the fraction"""
        return self.__num / self.__denom


# Do not change anything below.

f1 = Fraction(a, b)
f2 = Fraction(c, d)

print(f1)
print(f2)
print(f1 + f2)
print(f1 - f2)
print(f1 * f2)
print(f1 / f2)
f1.set_num(60)
print(f1)
print(f1.get_num(), f1.get_denom())

File: topics/class/fractions/att/template.cpp
#include <bits/stdc++.h>
using namespace std;

// A number represented as a fraction in reduced form.
// The denominator is always positive;
// the sign of the fraction is carried by the numerator.
class Fraction {
  private:
    int num;
    int denom;

    // Reduce fraction to lowest terms
    // Keep denominator positive
    void reduce() {
        // Write your code here
    }

  public:
    // num and denom!=0 are integers
    // call reduce here
    Fraction(int n, int d) {
        // Write your code here
    }

    // Getters
    int get_num() const { return num; }

    int get_denom() const { return denom; }

    // Setters
    void set_num(int value) {
        num = value;
        reduce();
    }

    void set_denom(int value) {
        assert(value != 0);

        denom = value;
        reduce();
    }

    double to_double() const { return (double)num / denom; }

    // Operator overloading
    // Returns a new fraction representing the addition
    Fraction operator+(const Fraction &other) const {
        // Write your code here
    }

    // Returns a new fraction representing the subtraction
    Fraction operator-(const Fraction &other) const {
        // Write your code here
    }

    // Returns a new fraction representing the multiplication
    Fraction operator*(const Fraction &other) const {
        // Write your code here
    }

    // Returns a new fraction representing the division
    Fraction operator/(const Fraction &other) const {
        // Write your code here
    }

    // Returns a string representation of this fraction
    // For example '-1/2' if numerator = -1 and denominator = 2
    friend ostream &operator<<(ostream &os, const Fraction &f) {
        // Write your code here
    }
};

// Do not change anything below.
int main() {
    int a, b, c, d;
    cin >> a >> b >> c >> d;

    Fraction f1(a, b);
    Fraction f2(c, d);

    cout << f1 << endl;
    cout << f2 << endl;
    cout << f1 + f2 << endl;
    cout << f1 - f2 << endl;
    cout << f1 * f2 << endl;
    cout << f1 / f2 << endl;

    f1.set_num(60);

    cout << f1 << endl;
    cout << f1.get_num() << " " << f1.get_denom() << endl;

    return 0;
}
